src/commands/application/instances/connect.rs

```diff
@@ -258,34 +309,175 @@ pub async fn handle_connect(context: Context, name: &str, port: &u16) -> Result<
-                instance_name.to_string(),
-            ));
+async fn cleanup_previous_livebook_instance(
+    client: &QueryClient,
+    application: &str,
+    namespace: &str,
+    version: &str,
+    preparing_progress_bar: indicatif::ProgressBar,
+) -> Result<(), CliError> {
+    let livebook_resource = client
+        .livebook_resource(application, namespace, version)
+        .await
+        .unwrap()
+        .data
+        .unwrap()
+        .livebook_resource;
+
+    let has_existing_livebook_pod = livebook_resource.is_some();
+
+    if has_existing_livebook_pod {
+        preparing_progress_bar.set_message("Found a provisioned Livebook instance belonging to you, re-creating your remote instance...");
+
+        debug!("Destroying the existing livebook instance.");
+        match client
+            .destroy_livebook(application, namespace, version)
+            .await
+        {
+            Ok(_) => {}
+            Err(err) => match &err {
+                crate::error::APIError::ResponseError { code, message } => {
+                    if !message.contains("pod_not_found") && !code.contains("pod_not_found") {
+                        return Err(err.into());
+                    }
+                }
+                _ => return Err(err.into()),
+            },
+        }
+
+        for i in 0..MAX_CHECKING_RETRY {
+            sleep(std::time::Duration::from_secs(RETRY_WAIT_TIME_IN_SEC)).await;
+
+            let livebook_resource = client
+                .livebook_resource(application, namespace, version)
+                .await?
+                .data
+                .unwrap()
+                .livebook_resource;
+
+            if livebook_resource.is_none() {
+                break;
+            }
+
+            if i == MAX_CHECKING_RETRY - 1 {
+                return Err(CliError::Timeout);
+            }
         }
     }
 
-    Err(CliError::InvalidInput {
-        value: name.to_string(),
-    })
+    Ok(())
 }
 
-async fn has_permission(
+async fn get_ready_k8s_pods(
     client: &QueryClient,
     application: &str,
     namespace: &str,
     version: &str,
-) -> Result<bool, CliError> {
-    Ok(client
-        .fetch_is_authorized(application, namespace, version)
+) -> Result<Vec<KubernetesPod>, CliError> {
+    let k8s_pods = client
+        .fetch_kubernetes_pods(application, namespace, version)
         .await?
         .data
         .unwrap()
-        .is_authorized)
+        .kubernetes_pods;
+
+    // filter out the pods that are not ready and livebook pods
+    let ready_pods = k8s_pods
+        .into_iter()
+        .map(|pod| KubernetesPod {
+            name: pod.name,
+            ready: pod.ready,
+            is_livebook: Some(pod.labels.contains(&"livebook".to_string())),
+        })
+        .filter(|pod| pod.ready && !pod.is_livebook.unwrap_or_default())
+        .collect();
+
+    Ok(ready_pods)
+}
+
+fn select_deployment_namespace() -> Result<Option<String>, CliError> {
```

Comment:
Question, why the return value need to be `Option`? Is it possible to get the index outside of the selection?

Reply:
Its not because of that but when I moved the selection to a function its expecting to have with default state which is null when user uses the ctr+c